src/parser/primitive.rs

```diff
@@ -0,0 +1,325 @@
+use std::io::Read;
+use std::mem::MaybeUninit;
+
+use byteorder::{BigEndian, ReadBytesExt};
+use num_traits::{FromPrimitive, ToPrimitive};
+
+use super::Error;
+
+#[allow(dead_code)]
+pub const ALIGNMENT: usize = 4;
+
+#[allow(dead_code)]
+fn read_padding(src: &mut dyn Read, n: usize) -> Result<(), Error> {
+    let mut buf = [0u8; ALIGNMENT];
+    let padding = (ALIGNMENT - n % ALIGNMENT) % ALIGNMENT;
+    src.read_exact(&mut buf[..padding]).map_err(|_| Error::IncorrectPadding)
+}
+
+#[allow(dead_code)]
+pub fn parse_u8(src: &mut dyn Read) -> Result<u8, Error> {
+    src.read_u8().map_err(Error::IO)
+}
+
+#[allow(dead_code)]
+pub fn parse_u32(src: &mut dyn Read) -> Result<u32, Error> {
+    src.read_u32::<BigEndian>().map_err(Error::IO)
+}
+
+#[allow(dead_code)]
+pub fn parse_u64(src: &mut dyn Read) -> Result<u64, Error> {
+    src.read_u64::<BigEndian>().map_err(Error::IO)
+}
+
+#[allow(dead_code)]
+pub fn parse_bool(src: &mut dyn Read) -> Result<bool, Error> {
+    let discr = parse_u32(src)?;
+    match discr {
+        0 => Ok(false),
+        1 => Ok(true),
+        _ => Err(Error::EnumDiscMismatch),
+    }
+}
+
+#[allow(dead_code)]
+pub fn parse_option<T>(
+    src: &mut dyn Read,
+    cont: impl FnOnce(&mut dyn Read) -> Result<T, Error>,
+) -> Result<Option<T>, Error> {
+    let disc = parse_bool(src)?;
+    match disc {
+        true => Ok(Some(cont(src)?)),
+        false => Ok(None),
+    }
+}
+
+#[allow(dead_code)]
+pub fn parse_array<const N: usize, T>(
+    src: &mut dyn Read,
+    cont: impl Fn(&mut dyn Read) -> Result<T, Error>,
+) -> Result<[T; N], Error> {
+    let mut buf: [MaybeUninit<T>; N] = [const { MaybeUninit::uninit() }; N];
+    for elem in buf.iter_mut() {
+        elem.write(cont(src)?);
+    }
+    read_padding(src, N * size_of::<T>())?;
+    Ok(unsafe { buf.as_ptr().cast::<[T; N]>().read() })
+}
+
+#[allow(dead_code)]
+pub fn parse_vector<T>(
+    src: &mut dyn Read,
+    cont: impl Fn(&mut dyn Read) -> Result<T, Error>,
+) -> Result<Vec<T>, Error> {
+    let size = parse_u32_as_usize(src)?;
+    let mut vec = Vec::with_capacity(size);
+    for _ in 0..size {
+        vec.push(cont(src)?);
+    }
+    read_padding(src, size_of::<T>() * size)?;
+    Ok(vec)
+}
+
+#[allow(dead_code)]
+pub fn parse_vec_max_size<const N: usize, T>(
```

Comment:
Not as const, just parameter.

For every const generic we generate new function which leads as to code bloat and reduced performanc